contracts/SpritzPayV3.sol

```diff
@@ -0,0 +1,291 @@
+                deadline,
+                path
+            )
+        );
+
+        emit Payment(
+            _paymentRecipient,
+            from,
+            sourceToken,
+            sourceTokenAmountSpent,
+            paymentToken,
+            paymentTokenAmount,
+            paymentReference
+        );
+    }
+
+    /**
+     * @notice Pay by swapping native currency. Uses
+     *          Uniswap exact output trade type
+     * @param path Swap path
+     * @param sourceTokenAmountMax Maximum amount of the token being sold for payment
+     * @param paymentTokenAmount Exact Amount of the target payment token
+     * @param paymentReference Arbitrary reference ID of the related payment
+     * @param deadline Swap deadline
+     */
+    function payWithNativeSwap(
+        uint256 sourceTokenAmountMax,
+        uint256 paymentTokenAmount,
+        bytes32 paymentReference,
+        uint256 deadline,
+        bytes calldata path
+    ) external payable whenNotPaused {
+        (address sourceToken, address paymentToken) = _swapModule.decodeSwapData(path);
+        if (!isAcceptedToken(paymentToken)) revert NonAcceptedToken(paymentToken);
+
+        uint256 sourceTokenAmountSpent = _swapModule.exactOutputNativeSwap{ value: sourceTokenAmountMax }(
+            SwapModule.ExactOutputParams({
+                to: _paymentRecipient,
+                from: msg.sender,
+                inputTokenAmountMax: sourceTokenAmountMax,
+                paymentTokenAmount: paymentTokenAmount,
+                deadline: deadline,
+                swapData: path
+            })
+        );
+
+        emit Payment(
+            _paymentRecipient,
+            msg.sender,
+            sourceToken,
+            sourceTokenAmountSpent,
+            paymentToken,
+            paymentTokenAmount,
+            paymentReference
+        );
+    }
+
+    /*
+     * Admin functions
+     */
+    function pause() external onlyRole(PAUSER_ROLE) {
+        _pause();
+    }
+
+    function unpause() external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _unpause();
+    }
+
+    function setPaymentRecipient(address newPaymentRecipient) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _setPaymentRecipient(newPaymentRecipient);
+    }
+
+    function setSwapModule(address newSwapTarget) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _setSwapModule(newSwapTarget);
+    }
+
+    function addPaymentToken(address newToken) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _addPaymentToken(newToken);
+    }
+
+    function removePaymentToken(address newToken) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _removePaymentToken(newToken);
+    }
+
+    function revokePaymentDelegate(address delegate) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _revokeRole(PAYMENT_DELEGATE_ROLE, delegate);
+    }
+
+    function grantPaymentDelegate(address delegate) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _grantRole(PAYMENT_DELEGATE_ROLE, delegate);
+    }
```

Comment:
do we need these 2? I thought the basic grantRole /revokeRole functions are already available to us without a particular interface

Reply:
good catch!